// src/services/mockData.js
// src/services/mockData.js

// --- CUSTOMER MOCK DATA ---
export const mockUser = {
  name: "Mohammed Amine",
  firstName: "Mohammed",
  lastName: "Amine",
  clientId: "789456", 
  email: "mohammed.amine@email.com",
  phone: "+212 612 345 678",
  address: "123 Boulevard Mohammed, Algiers, Algeria",
};

export const mockProblems = [
  { id: "#P345892", title: "La TV ne s'allume pas", signalDate: "20 mars 2025", status: "En cours", statusKey: "inprogress", details: "Le technicien a diagnostiqué un problème d'alimentation et attend la pièce.", product: "Stream TV 55\"", productSerial: "BOM-TV55-234567", lastUpdate: "21 mars 2025 - Pièce commandée, ETA 3 jours.", assignedTechnician: "Ahmed K."},
  { id: "#P346001", title: "Problème de télécommande", signalDate: "22 mars 2025", status: "En attente de réponse", statusKey: "pending", details: "Le client doit fournir le modèle exact de la télécommande.", product: "Stream TV 43\"", productSerial: "PROD_TV43_XYZ", lastUpdate: "22 mars 2025 - Information complémentaire demandée au client.", assignedTechnician: "Sophie L."},
  { id: "#P344920", title: "Problème de connectivité WIFI", signalDate: "15 mars 2025", status: "Résolu", statusKey: "resolved", details: "Problème résolu avec une mise à jour du firmware appliquée par le client.", product: "Stream TV 32\"", productSerial: "BOM-TV32-123456", lastUpdate: "17 mars 2025 - Client confirme résolution.", assignedTechnician: "Client (Auto-résolu)"},
  { id: "#P345100", title: "Remplacement d'écran Smartphone", signalDate: "17 mars 2025", status: "Planifié", statusKey: "planned", details: "Intervention planifiée pour le 25 mars 2025 à 14h00.", product: "Smartphone X5", productSerial: "BOM-SPX5-987654", lastUpdate: "18 mars 2025 - Rendez-vous confirmé avec le client.", assignedTechnician: "Fatima Z."},
];

export const mockProducts = [
  { id: "prod1", name: 'Stream TV 55" OLED', type: 'tv', serial: "BOM-TV55-234567", purchaseDate: "15 janvier 2024", warrantyExpiration: "15 janvier 2026", status: "Active", statusKey: "active", specifications: { resolution: "4K UHD", smart_features: "Android TV, Google Assistant", ports: "3x HDMI, 2x USB, Ethernet" } },
  { id: "prod2", name: "Smartphone X5 Pro", type: 'smartphone', serial: "BOM-SPX5-987654", purchaseDate: "10 mars 2024", warrantyExpiration: "10 mars 2026", status: "Active", statusKey: "active", specifications: { os: "Android 13", camera: "108MP Triple Camera", storage: "256GB" } },
  { id: "prod3", name: 'Stream TV 32" HD', type: 'tv', serial: "BOM-TV32-123456", purchaseDate: "5 février 2022", warrantyExpiration: "5 février 2024", status: "Expiré", statusKey: "expired", specifications: { resolution: "HD Ready", ports: "2x HDMI, 1x USB" } },
  { id: "prod4", name: 'Soundbar BOMARE SB200', type: 'audio', serial: "BOM-AUD-SB200-001", purchaseDate: "1 juin 2023", warrantyExpiration: "1 juin 2025", status: "Active", statusKey: "active", specifications: { channels: "2.1", connectivity: "Bluetooth, Optical", power: "120W" } }
];

export const mockKnowledgeBase = [
  { id: 'kb001', title: "Comment mettre à jour le firmware de votre TV BOMARE Série X", date: "15/03/2025", category: "TVs", content_summary: "Ce guide vous explique..." },
  { id: 'kb002', title: "Instructions de réinitialisation d'usine pour les smartphones BOMARE X", date: "18/03/2025", category: "Smartphones", content_summary: "Suivez ces étapes pour..." },
  { id: 'kb003', title: "Connexion de votre TV intelligente BOMARE à un réseau domestique Wi-Fi", date: "20/03/2025", category: "TVs", content_summary: "Difficultés de connexion? Vérifiez..." },
  { id: 'kb004', title: "Optimiser la qualité sonore de votre Soundbar BOMARE", date: "10/02/2025", category: "Audio domestique", content_summary: "Pour une meilleure expérience..." },
];

export const mockNotifications = [
  { id: 1, type: 'info', title: "Informations Supplémentaires Requises pour #P346001", message: "Un technicien a demandé plus d'informations...", time: "Hier, 15:45", unread: true, link: "/customer/my-problems/respond/#P346001" },
  { id: 2, type: 'assignment', title: "Technicien Assigné à Votre Problème #P345892", message: "Ahmed K. a été assigné...", time: "Aujourd'hui, 10:30", unread: true, link: "/customer/my-problems/detail/#P345892" },
  { id: 3, type: 'update', title: "Mise à jour du Problème #P345892", message: "Le statut de votre problème est passé à 'En cours - Pièce en attente'.", time: "Aujourd'hui, 11:15", unread: false, link: "/customer/my-problems/detail/#P345892" },
  { id: 4, type: 'warranty', title: "Rappel: Garantie Expirant Bientôt", message: "La garantie de votre Smartphone X5 (BOM-SPX5-987654) expirera dans 30 jours.", time: "Il y a 2 jours", unread: true, link: "/customer/warranty-products/detail/prod2" },
  { id: 5, type: 'system', title: "Maintenance Programmée du Portail", message: "Une maintenance du portail client est prévue le 30 mars 2025 de 02:00 à 04:00.", time: "Il y a 1 semaine", unread: false, link: "#" }
];

// --- DATA FROM YOUR FRIEND (TECHNICIAN RELATED) ---
// Ensuring this array is exported with the name 'videos'
export const videos = [
    {
        id: 'vid001', // Changed to string ID like other mocks for consistency
        title: 'Easy Steps to Fix Your TV Tech Revealed',
        thumbnail: 'https://via.placeholder.com/400x225/1E3A5F/FFFFFF?text=TV+Repair+101',
        duration: '45:12', // More standard duration format
        level: 'Débutant', // Using French for levels
        type: 'tv', // Product type for filtering
        technician: 'Adrian Williams (BOMARE Expert)',
        description: 'Apprenez les étapes de dépannage de base pour les problèmes de télévision courants avec Adrian. Couvre les modèles communs de BOMARE.',
        uploadDate: '2024-03-10T09:30:00Z', // ISO date string for easier sorting
        videoUrl: 'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4', 
        tags: ['TV TECH', 'Dépannage', 'Réparation', 'Diagnostic'],
        views: 1250,
        likes: 130,
        dislikes: 5,
        technicianAvatar: '/assets/images/avatar-tech-male1.png', // Placeholder path
    },
    {
        id: 'vid002',
        title: 'Troubleshoot Any TV That Doesn\'t Turn On',
        thumbnail: 'https://via.placeholder.com/400x225/3F2E56/FFFFFF?text=TV+Power+Issues',
        duration: '30:55',
        level: 'Intermédiaire',
        type: 'tv',
        technician: 'Kendall Toole (BOMARE Pro)',
        description: 'Session ciblée sur le diagnostic des problèmes liés à l\'alimentation des téléviseurs, spécifiquement pour les séries X et Y.',
        uploadDate: '2024-04-01T14:00:00Z',
        videoUrl: 'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ElephantsDream.mp4',
        tags: ['HOW TO', 'Alimentation', 'Panne TV', 'Electronique'],
        views: 2340,
        likes: 250,
        dislikes: 8,
        technicianAvatar: '/assets/images/avatar-tech-female1.png',
    },
    {
        id: 'vid003',
        title: 'Calibration Audio pour Home Cinéma BOMARE', // Made title more relevant
        thumbnail: 'https://via.placeholder.com/400x225/4B78BD/FFFFFF?text=Audio+Calibration',
        duration: '15:20',
        level: 'Avancé',
        type: 'audio', // Product type
        technician: 'Emma Lovewell (Ingénieur Son)',
        description: 'Guide avancé pour la calibration optimale des systèmes audio home cinéma BOMARE pour une expérience immersive.',
        uploadDate: '2024-05-21T11:00:00Z',
        videoUrl: 'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4',
        tags: ['Audio', 'Calibration', 'Home Cinéma', 'Expert'],
        views: 1890,
        likes: 210,
        dislikes: 3,
        technicianAvatar: '/assets/images/avatar-tech-female2.png',
    },
    // Example for an "article" type content
    {
        id: 'article001',
        title: 'Guide de Dépannage pour les Problèmes Communs des TVs BOMARE',
        thumbnail: 'https://via.placeholder.com/400x225/1c1d42/c0c0e0?text=Guide+Article',
        duration: 'Article', 
        level: 'Tous les Niveaux',
        type: 'article', 
        technician: 'Documentation BOMARE',
        description: 'Consultez notre guide écrit complet pour résoudre les problèmes les plus fréquents rencontrés sur nos téléviseurs.',
        uploadDate: '2024-01-10T10:00:00Z',
        videoUrl: null, // Signifies it's not a video
        externalLink: 'https://www.bomare.dz/support/tv-guides', // Example actual link
        tags: ['Guide', 'Dépannage', 'Texte', 'FAQ Link'],
        views: 5050, likes: 0, dislikes: 0,
        technicianAvatar: '/assets/images/bomare-logo.png', // Or a generic BOMARE logo
    },
];

// Technician User Profile Mock Data (from your friend's 'userProfile', adapted)
export const mockTechnicianUser = {
  id: 'tech_john_doe_001', // Unique ID for the technician
  name: 'John Doe', 
  firstName: 'John',    // Added for consistency with mockUser if needed elsewhere
  lastName: 'Doe',
  technicianId: 'TECH12345', // From her data
  email: 'john.doe@bomare.com',
  phone: '+213 770 123 456', // Example Algerian format
  avatarUrl: 'https://via.placeholder.com/150/1E3A5F/FFFFFF?text=JD', // Using a placeholder for now
  role: 'technician',     // Important for role-based logic in App.js
  tagline: 'Technicien Senior | Expert en Solutions Électroniques',
  isOnline: true,
  bio: 'Technicien certifié avec plus de 8 ans d\'expérience. Spécialisé dans le diagnostic des systèmes TV intelligents et la réparation de composants. Engagé à fournir un service rapide et efficace.',
  location: 'Alger, Algérie',
  joinDate: 'Février 2018',
  skills: ['Réparation TV OLED/LED', 'Diagnostic de Pannes Complexes', 'Calibration Audio/Vidéo', 'Firmware BOMARE OS', 'Service Clientèle Technique'],
  connections: [
    { id: 'conn1', name: 'Amina K.', role: 'Technicienne Support N2', avatarUrl: 'https://via.placeholder.com/100/4ADE80/0E1330?text=AK' },
    { id: 'conn2', name: 'Youssef B.', role: 'Spécialiste Pièces Détachées', avatarUrl: 'https://via.placeholder.com/100/E9655A/FFFFFF?text=YB' },
  ],
  emailNotifications: true,
  darkMode: false,
  watchedVideos: ['video1', 'vid002'] // IDs of videos this technician has watched
};
